packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -61,40 +63,95 @@ export interface IBlobManagerLoadInfo {
+        30 * 1000, // 30 sec max delay
+        // throttling function increases exponentially (0ms, 40ms, 80ms, 160ms, etc)
+        formExponentialFn({ coefficient: 20, initialDelay: 0 }),
+    );
 
     constructor(
         private readonly routeContext: IFluidHandleContext,
         snapshot: IBlobManagerLoadInfo,
         private readonly getStorage: () => IDocumentStorageService,
-        private readonly attachBlobCallback: (blobId: string) => void,
+        private readonly sendBlobAttachOp: (blobId?: string, localId?: string) => void,
         // To be called when a blob node is requested. blobPath is the path of the blob's node in GC's graph. It's
         // of the format `/<BlobManager.basePath>/<blobId>`.
         private readonly gcNodeUpdated: (blobPath: string) => void,
-        private readonly runtime: IContainerRuntime,
-        private readonly logger: ITelemetryLogger,
+        private readonly runtime: IBlobManagerRuntime,
     ) {
-        this.runtime.once("dispose", () => {
-            for (const promise of this.pendingBlobIds.values()) {
-                promise.reject(new Error("runtime disposed while blobAttach op in flight"));
-            }
-        });
-        this.load(snapshot);
+        this.logger = ChildLogger.create(this.runtime.logger, "BlobManager");
+        this.runtime.on("disconnected", () => this.onDisconnected());
+        this.redirectTable = this.load(snapshot);
     }
 
-    private hasBlob(id: string): boolean {
-        return this.blobIds.has(id) || this.detachedBlobIds.has(id);
+    public get pendingOfflineUploadCount(): number {
+        return Array.from(this.pendingBlobs.values())
+            .filter((e) => e.status === PendingBlobStatus.OfflinePendingUpload).length;
+    }
+
+    public get hasPendingOfflineUploads(): boolean {
+        return this.pendingOfflineUploadCount > 0;
+    }
+
+    /**
+     * Upload blobs added while offline. This must be completed before connecting and resubmitting ops.
+     */
+    public async onConnected() {
+        await PerformanceEvent.timedExecAsync(this.logger, {
+                eventName: "BlobUploadOnConnected",
+                count: this.pendingOfflineUploadCount,
+            }, async () => Promise.all(Array.from(this.pendingBlobs.values())
+                    .filter((e) => e.status === PendingBlobStatus.OfflinePendingUpload).map(async (e) => e.uploadP)),
+            { start: true, end: true },
+        );
+    }
+
+    private onDisconnected() {
+        for (const [localId, entry] of this.pendingBlobs) {
+            if (entry.status === PendingBlobStatus.OnlinePendingOp) {
+                // we won't see the op until we connect again
+                entry.status = PendingBlobStatus.OfflinePendingOp;
+                // we submitted an op with the storage ID already, but it didn't
+                // have the local ID and we may not get a chance to resubmit it
+                this.sendBlobAttachOp(entry.storageId, localId);
```

Comment:
Hmm...  I was expecting we would have a special case to handle blobs that haven't yet finished uploading, but if we have the storage id already, doesn't that mean the blob upload has finished and we just need to send a blobAttach op to announce it?  I'm not sure why we fall back on localIds & redirectTable in that case.

...I suspect I have something mixed up, but I'm not sure what it is.  Can you tell where I'm confused?

Reply:
> I was expecting we would have a special case to handle blobs that haven't yet finished uploading,

Vlad suggested (if I understood correctly) not handling this on disconnect and just letting the upload call resolve/reject. If the client is offline we expect it to reject eventually.

> if we have the storage id already, doesn't that mean the blob upload has finished and we just need to send a blobAttach op to announce it? I'm not sure why we fall back on localIds & redirectTable in that case.

Yes. In general we still wait for the BlobAttach op to round-trip before returning the handle because it's possible the BlobAttach op didn't go through successfully, then pending changes including the blob are stashed, then reapplied later when the blob is gone from storage. In this case, it's possible we won't get the same ID from storage when we re-upload, so the handle is broken. If we assume we will always get the same ID then I think it's safe to return early.

Here we need to resolve the handle on disconnect event since if it's caused by the client going offline, the op will not round-trip until they go back online, and we need to support offline work.